finality-aleph/src/validator_network/io.rs

```diff
@@ -0,0 +1,176 @@
+) -> Result<(S, D), ReceiveError> {
+    let mut buf = [0; 4];
+    stream
+        .read_exact(&mut buf[..])
+        .await
+        .map_err(|_| Error::ConnectionClosed)?;
+    let len = u32::from_le_bytes(buf);
+    if len > MAX_DATA_SIZE {
+        return Err(Error::DataTooLong(len).into());
+    }
+    let mut buf: Vec<u8> = vec![0; len as usize];
+    stream
+        .read_exact(&mut buf[..])
+        .await
+        .map_err(|_| Error::ConnectionClosed)?;
+    let data = D::decode(&mut &buf[..]).map_err(|_| ReceiveError::DataCorrupted)?;
+    Ok((stream, data))
+}
+
+#[cfg(test)]
+mod tests {
+    use tokio::io::{duplex, AsyncWriteExt};
+
+    use super::{receive_data, send_data, Error, ReceiveError, SendError, MAX_DATA_SIZE};
+
+    #[tokio::test]
+    async fn sends_and_receives_correct_data() {
+        let (sender, receiver) = duplex(4096);
+        let data: Vec<i32> = vec![4, 3, 43];
+        let _sender = send_data(sender, data.clone())
+            .await
+            .expect("data should send");
+        let (_receiver, received_data) = receive_data(receiver).await.expect("should receive data");
+        let received_data: Vec<i32> = received_data;
+        assert_eq!(data, received_data);
+    }
+
+    #[tokio::test]
+    async fn fails_to_receive_from_dropped_connection() {
+        let (_, receiver) = duplex(4096);
+        match receive_data::<_, i32>(receiver).await {
+            Err(e) => assert_eq!(ReceiveError::Error(Error::ConnectionClosed), e),
+            _ => panic!("received data from a dropped stream!"),
+        }
+    }
+
+    #[tokio::test]
+    async fn fails_to_send_to_dropped_connection() {
+        let (sender, _) = duplex(4096);
+        let data: Vec<i32> = vec![4, 3, 43];
+        match send_data(sender, data.clone()).await {
+            Err(e) => assert_eq!(SendError(Error::ConnectionClosed), e),
+            _ => panic!("send data to a dropped stream!"),
+        }
+    }
+
+    #[tokio::test]
+    async fn fails_to_receive_too_much_data() {
+        let (mut sender, receiver) = duplex(4096);
+        let too_long = MAX_DATA_SIZE + 43;
+        let payload = too_long.to_le_bytes().to_vec();
+        sender
+            .write_all(&payload)
+            .await
+            .expect("sending should work");
+        match receive_data::<_, i32>(receiver).await {
+            Err(e) => assert_eq!(ReceiveError::Error(Error::DataTooLong(too_long)), e),
+            _ => panic!("received too long data!"),
+        }
+    }
+
+    #[tokio::test]
+    async fn fails_to_decode_empty_data() {
+        let (mut sender, receiver) = duplex(4096);
+        let payload = 0u32.to_le_bytes().to_vec();
+        sender
+            .write_all(&payload)
+            .await
+            .expect("sending should work");
+        match receive_data::<_, i32>(receiver).await {
+            Err(e) => assert_eq!(ReceiveError::DataCorrupted, e),
+            _ => panic!("decoded no data into something?!"),
+        }
+    }
+}
```

Comment:
We fail early, so if we get bad data once we drop the connection and thus the latter situation doesn't need to be tested.

The former I did not want to test as it requires allocating 16MB of memory for no reason. :< Should we?

Reply:
It's for free :) And writing such test can also provide some improvements, i.e. you can consider calling `size_hint` first instead of trying to allocate potentially too big message first. It would also justify usage of `codec` here.

Reply:
Added test, but I did not do the thing with `size_hint` – it would result in strangely repeated code, and allocating a big vector for encoding implies we already allocated a lot of space for the message itself, plus it will never happen in practice (if we actually cannot send a message we created we are in far bigger trouble).